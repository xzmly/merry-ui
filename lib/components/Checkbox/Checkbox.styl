@import '../../helpers/stylus.helper.styl';

checkbox = class-prefix + "-checkbox"
input = checkbox + "-input"
input-wrap = input + "-wrap"
input-inner = input + "-inner"
input-text = checkbox + "-text"


.{input-wrap}
  margin 0
  padding 0
  outline none
  position relative
  display inline-block
  box-sizing border-box
  white-space nowrap
  vertical-align middle
  list-style: none
  width 16px
  height 16px
  overflow hidden
  top: -0.09em
  font-size 14px


.{input}
  position absolute
  top 0
  right 0
  bottom 0
  left 0
  width 100%
  height 100%
  z-index 1
  margin 0
  padding 0
  border 0
  cursor pointer
  box-sizing border-box
  opacity 0
  &:checked + span.{input-inner}
    background primary-color
    &:after
      content: " "
      position absolute
      width: 5px;
      height: 12px;
      transform rotate(45deg) scale(0.7) translate(-50%,-50%)
      border  2px solid #fff
      border-left 0
      border-top 0
      top 28%
      left 17%


.{input-inner}
  position relative
  display inline-block
  width 16px
  height 16px
  border: 1px solid #d9d9d9;
  border-radius 2px
  border-collapse: separate;
  transition all .3s
  vertical-align center
  box-sizing border-box

.{input-text}
  padding 0 8px
  font-size 14px
  color rgba(0,0,0,0.75)
  cursor pointer

