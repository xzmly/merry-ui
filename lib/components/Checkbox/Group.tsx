import * as React from 'react';
import {useState, useEffect} from "react"
import classes from '../../helpers/classes';
import Checkbox, {onChangeType} from "./Checkbox";

type ValueType = string | number | boolean

export interface OptionType {
  label: React.ReactNode;
  value: ValueType;
  disabled?: boolean;
  onChange?: onChangeType;
}

export interface GroupProps {
  defaultValue?: Array<ValueType>
  disabled?: boolean

  options?: Array<OptionType>
  value?: Array<ValueType>
  onChange?: (checkedValue: Array<ValueType>) => void
}

const Group: React.FC<GroupProps> =
  props => {

    const [defaultValue, setDefaultValue] =
        useState<Array<ValueType> | undefined>(props.defaultValue);

    const {
      value,
      options,
      disabled,
    } = props;

    useEffect(() => {
      if (value) return;
      setDefaultValue(props.defaultValue);
    }, [props.defaultValue]);

    const valueOrDefaultValue: Array<ValueType> = value || defaultValue || [];

    const onChange = (
        event: React.ChangeEvent<HTMLInputElement>,
        fn?: onChangeType
    ): void => {
      const {checked, value} = event.target;

      const newValue: Array<ValueType> = checked ?
          [...valueOrDefaultValue, value] :
          valueOrDefaultValue.filter(v => v !== value);

      !props.value && setDefaultValue(newValue);
      fn && fn(event);
      props.onChange && props.onChange(newValue);
    };

    const renderChildren: React.ReactNodeArray | undefined =
        options?.map((v: OptionType) =>
            <Checkbox
                checked={valueOrDefaultValue.includes(v.value)}
                onChange={(event) => onChange(event, v.onChange)}
                disabled={'disabled' in v ? v.disabled : disabled}
                key={v.value.toString()}
                value={v.value.toString()}
            >
              {v.label}
            </Checkbox>);

    return (
        <div className={classes('checkbox-group')}>
          {renderChildren}
        </div>
    )
  };

export default Group