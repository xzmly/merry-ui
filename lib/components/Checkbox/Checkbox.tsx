import * as React from 'react';
import classes from '../../helpers/classes';
import Group, { GroupProps } from "./Group"
import './Checkbox.styl';

export interface LabelProps {
  className?: string
  children?: React.ReactNode
  htmlFor?: string
  form?: string
  style?: React.CSSProperties
}

export type onChangeType = (event:React.ChangeEvent<HTMLInputElement>) => void

export interface InputCheckboxProps extends React.InputHTMLAttributes<HTMLInputElement>{}

interface CheckboxComponent<P> extends React.FC<P>{
  Group: React.FC<GroupProps>
}

const Checkbox: CheckboxComponent<LabelProps & InputCheckboxProps> =
        props => {

  const { 
    className,
    children,
    htmlFor,
    form,
    style,
    ...checkboxProps
  } = props;

  const restClass = (name:string):string =>
      classes('',`checkbox-${name}`);
  
  const onChange:onChangeType = (event):void => {
    checkboxProps.onChange && checkboxProps.onChange(event);
  };

  return(
      <label className={classes(
          className,
          'checkbox',
          checkboxProps.disabled ? 'checkbox-disabled' : ''
      )}
             htmlFor={htmlFor}
             form={form}
             style={style}
      >
        <span className={restClass('input-wrap')}>
          <input {...checkboxProps}
                 type="checkbox"
                 onChange={onChange}
                 className={restClass('input')}
          />
           <span className={restClass('input-inner')}/>
        </span>
        <span className={restClass(`text`)}>
          {children}
        </span>
      </label>
  )
};

Checkbox.Group = Group;

export default Checkbox