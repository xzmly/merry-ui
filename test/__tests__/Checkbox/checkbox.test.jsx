import React from 'react'
import Checkbox from '../../../lib/components/Checkbox/Checkbox'
import renderer from "react-test-renderer";
import {mount} from "enzyme";

const { Group } = Checkbox;

describe('Checkbox',()=>{
  it('render Checkbox',()=>{
    const json = renderer.create(<Checkbox/>).toJSON();
    expect(json).toMatchSnapshot()
  });
  it('render Group',()=>{
    const json = renderer.create(<Group/>).toJSON();
    expect(json).toMatchSnapshot()
  });
  it('受控组件',()=>{
    const fn = jest.fn();
    const c = mount(<Checkbox checked={true} onChange={fn}/>);
    expect(c.find('.merry-checkbox-input').props().checked).toBe(true);
    c.setProps({ checked: false });
    expect(c.find('.merry-checkbox-input').props().checked).toBe(false);
    c.find('.merry-checkbox-input').simulate('change');
    expect(fn).toHaveBeenCalledTimes(1)
  });
  it('非受控组件',()=>{
    const fn = jest.fn();
    const c = mount(<Checkbox defaultChecked={true} onChange={fn}/>);
    c.setProps({ defaultChecked: false });
    expect(c.find('.merry-checkbox-input').props().defaultChecked).toBe(false);
    c.find('.merry-checkbox-input').simulate('change');
    expect(fn).toHaveBeenCalledTimes(1)
  });
  it('disabled',()=>{
    const c = mount(<Checkbox disabled/>);
    c.find('.merry-checkbox-input').props().disabled;
    expect(c.find('.merry-checkbox-input').props().disabled).toBe(true)
  })
});
