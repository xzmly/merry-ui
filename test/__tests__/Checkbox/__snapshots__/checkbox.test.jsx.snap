// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox render Checkbox 1`] = `
<label
  className="merry-checkbox"
>
  <span
    className="merry-checkbox-input-wrap"
  >
    <input
      className="merry-checkbox-input"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="merry-checkbox-input-inner"
    />
  </span>
  <span
    className="merry-checkbox-text"
  />
</label>
`;

exports[`Group render Group 1`] = `
<div
  className="checkbox-group"
/>
`;
